KAFKA-17248 - KIP 1076 implementation

bbejeck opened the pull request:
Implementation of KIP-1076 to allow for adding client application metrics to the KIP-714 framework

Integration and unit tests forthcoming


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

bbejeck committed:
Fix merge conflicts from breaking into smaller PR

bbejeck committed:
Resolving merge conflics

bbejeck committed:
Fixing rebase

bbejeck committed:
Fix producer javadoc comment
Correct metric filtering to only include metrics with thread-id
Remove unused method

bbejeck committed:
Fix spotbugs and checkstyle errors

bbejeck committed:
Simplify filtering for adding metrics

bbejeck committed:
Clean up the creation of the reporter such that it relys on the metrics infra totally for adding removing metrics.

bbejeck committed:
Resolving merge conflict from this previous commit Added parameterized integration test

bbejeck committed:
Added test for passing streams metrics

bbejeck committed:
Added integration tests, set logging level in reporter to debug

bbejeck committed:
Added test to confirm stream metrics don't bleed into consumer metrics

bbejeck committed:
Rebasing work

bbejeck committed:
Add some description to large test method

bbejeck committed:
checkstyle fixes

bbejeck committed:
Remove unintentional formatting

apoorvmittal10 commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -621,6 +631,9 @@ private KafkaAdminClient(AdminClientConfig config,
         this.clientTelemetryEnabled = config.getBoolean(AdminClientConfig.ENABLE_METRICS_PUSH_CONFIG);
+        List<MetricsReporter> reporters = CommonClientConfigs.metricsReporters(this.clientId, config);
+        this.clientTelemetryReporter = clientTelemetryReporter;
+        this.clientTelemetryReporter.ifPresent(reporters::add);
```
Should we add close of the client telemetry reporter on admin client close?

mjsax replied:
Should reporters not be closed when admin client calls `metrics.close()` (I would assume it does correctly close it's `metrics` object?)

Following up, because I don't see code changes about it yet.

bbejeck replied:
Added  `metrics.close()`to the `KafkaAdmin.close()` - `metrics.close()` does close all reporters

apoorvmittal10 replied:
Hmmm, but the close will not send the last telemetry push as ClientTelemetryReporter need `initiateClose` method call to set state as `terminating_push_needed` which sends the last telemetry metrics to broker. However if we decide not to send last telemetry push in admin client that's fine but we should write it in comments somewhere.

bbejeck replied:
@apoorvmittal10 good catch, I've updated the close method to call `initiateClose`, the use in the admin client should follow the same semantics of the other clients.

apoorvmittal10 replied:
👍

apoorvmittal10 commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -547,6 +551,8 @@ static KafkaAdminClient createInternal(
             MetricsContext metricsContext = new KafkaMetricsContext(JMX_PREFIX,
                     config.originalsWithPrefix(CommonClientConfigs.METRICS_CONTEXT_PREFIX));
             metrics = new Metrics(metricConfig, reporters, time, metricsContext);
+            clientTelemetryReporter = CommonClientConfigs.telemetryReporter(clientId, config);
+            clientTelemetryReporter.ifPresent(telemetryReporter -> telemetryReporter.contextChange(metricsContext));
```
Though it's good to have the integartion of KIP-714 with admin client but I left it last time because of adhoc connections by Admin client, this will establish telemetry connection to send metrics, do we require them? Secondly, does admin client has any metric of it's own in Kafka? At the time of KIP-714, I didn't see any relevant metric for admin client. Last month we added https://issues.apache.org/jira/browse/KAFKA-17239 to record node latency which was related to https://issues.apache.org/jira/browse/KAFKA-17230. But again should we have KIP-714 pipeline for this 1 admin metric, I am not sure.

Hence should we register additional metrics here? Seems more of an additional overhead for adhoc requests. Else if we want external applications to register additional metrics in Admin client anyways then we might want to consider only enabling client telemetry reporter if additional metrics are registered. Wdyt?

bbejeck replied:
That's a good point, I'll take a look at doing that.

mjsax replied:
> do we require them?

Yes, as we want to use `Admin` to report KS instance metrics.

> Secondly, does admin client has any metric of it's own in Kafka

I would assume so? Why would we create a `new Metrics` object if we don't use it? (Even if I have to admit, the docs don't say anything about admin metrics... 🤔 -- maybe @cmccabe can help on this question?)

> we might want to consider only enabling client telemetry reporter if additional metrics are registered. Wdyt?

I guess we could do this -- but it's kinda odd to have the Admin config "enable push metrics" but not enable it if set to `true`...? -- If we don't want to enable it on the admin client, we should rather remove (or for now only deprecate) the existing config, and add a new one via KIP-1076 which would enable "application push metrics" which would only create a reporter if application metrics are registered -- otherwise, we end up with some weird behavior which can easily confuse users.

Btw: I was just looking into the consumer/producer code, and it seem there we just create the reporter and add to `reporters` list before `new Metrics` is called -- this way, `metricsContext` will be "added" automatically to the reporter. Should simplify the code a little bit?

bbejeck replied:
I'm leaning toward not doing this, as I agree with @mjsax. Plus, there's additional complexity in the case where metrics are unregistered. Should that trigger the mirror action of disabling metrics pushing? 

>Btw: I was just looking into the consumer/producer code, and it seem there we just create the reporter and add to reporters list before new Metrics is called -- this way, metricsContext will be "added" automatically to the reporter. Should simplify the code a little bit?

I'm not sure I follow this comment.  Are you referring to the reporters we create in Kafka Streams to act as proxies for passing the metrics to the clients and suggesting we create them in the consumer/producer clients themselves?   
Those `Reporter` instances are specific to Kafka Streams only.  Also, the `Metrics` API provides an `addReporter` method, so it seems it was designed to add a reporter after instantiating the `Metrics` object.

apoorvmittal10 replied:
So the only metric I think admin client emits is `node latency` metric, which also got recently added as regression. Though I understand that KS wants to send application metrics as part of admin client but we will be adding an additional overhead with short lived admin client operations. I am not very aligned regarding with KIP-714 integration in admin client. @AndrewJSchofield wdyt?

bbejeck replied:
@apoorvmittal10 I spoke offline with @mjsax and what we've come up with is that we'll disable metrics in the AdminClient by default (I'll update KIP-1076 with this detail).  KafkaStreams will explicitly enable the metrics push for the AdminClient, avoiding any additional overhead for non-KafkaStreams usage.  WDYT?

apoorvmittal10 replied:
That sounds fair to me. For my knowledge, does admin client in streams application has a long running connection where metrics will be emitted? My concern is with if connections are adhoc and we do create too many admin client short lived connections then it might impact the connection cache i.e. connection cache keeps the client instance for maximum of push interval * 3 ms hence if too many admin client connections created in short span then it might hurt the performance.

AndrewJSchofield replied:
Sounds good to me too.

mjsax replied:
Kafka Streams uses the admin client to also send "delete record" request for repartition topics. For this case, the connection should be long lived, as we send a "delete record" request after each commit by default (it's actually configurable).

For the (more rare) case that there is no repartition topic, it could be a slightly different story, but IIRC, @bbejeck did some tests already, verifying that the connection is not closed by long lived, even for this case. So I think we are good?

We could run more test before the 4.0 release to double check, and if we really find an issue, we could still update the KIP, to let KS use different connection timeout default, to make sure that the connection will be long lived.

apoorvmittal10 commented on clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java:
```diff
@@ -71,6 +72,7 @@ public class MockProducer<K, V> implements Producer<K, V> {
     private boolean producerFenced;
     private boolean sentOffsets;
     private long commitCount = 0L;
+    final List<KafkaMetric> addedMetrics = new ArrayList<>();
```
Rather `default` should we mark this as private and expose a getter which should return unmodifiable copy of `addedMetrics`? Exposing getter will be as what other members in the class currently exposed as, unmodifiable copy will limit the scope of error in downstream tests by mistakenly modifying the list.

Also shall we have lazy initializtion for the member?

Same as above for other files.

apoorvmittal10 commented on clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java:
```diff
@@ -1484,4 +1485,14 @@ public synchronized List<Node> brokers() {
+
+    @Override
+    public void registerMetricForSubscription(KafkaMetric metric) {
+
+    }
+
+    @Override
+    public void unregisterMetricForSubscription(KafkaMetric metric) {
+
```
Is there a reason we do not want to do same functionality as like MockProducer and MockConsumer here? If there is then can we please write comments here.

bbejeck replied:
oversight on my part

apoorvmittal10 commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadDelegatingMetricsReporter.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.Objects;
+
+public class StreamsThreadDelegatingMetricsReporter implements MetricsReporter {
+
```
nit: Remove additional line break.

apoorvmittal10 commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadDelegatingMetricsReporter.java:
```diff
@@ -0,0 +1,87 @@
+    private Consumer<?, ?> consumer;
+    final String threadId;
+    final String stateUpdaterThreadId;
+    private static final String THREAD_ID_TAG = "thread-id";
```
nit (can be ignored): shall we structure constants above the instance members?

apoorvmittal10 commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadDelegatingMetricsReporter.java:
```diff
@@ -0,0 +1,87 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(StreamsThreadDelegatingMetricsReporter.class);
+    private Consumer<?, ?> consumer;
+    final String threadId;
+    final String stateUpdaterThreadId;
```
Should we mark them private and expose `threadId()` and `stateUpdaterThreadId()` if required?

bbejeck replied:
I've marked them as private, I can't recall why I didn't do that in the first place

apoorvmittal10 commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadDelegatingMetricsReporter.java:
```diff
@@ -0,0 +1,87 @@
+
+    @Override
+    public void close() {
+        this.consumer = null;
```
Should we call `consumer.close()` here before setting to null?

bbejeck replied:
Kafka Streams fully controls the consumer's life cycle, so calling close at this point would lead to unexpected behavior. Setting the consumer to `null` here prevents memory leaks by maintaining a reference to the embedded consumer. However, Kafka Streams does call `Metrics.close` on shutdown, which calls close on all reporters, so I suspect that won't be a problem - I'll leave as is for now.

EDIT: I've taken another look and decided to remove setting the consumer to null here, even though I don't think it would be an issue with `Metrics.close` being called after Kafka Streams closes the consumer, better safe than sorry.

apoorvmittal10 commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadDelegatingMetricsReporter.java:
```diff
@@ -0,0 +1,87 @@
+public class StreamsThreadDelegatingMetricsReporter implements MetricsReporter {
+
+
+    private static final Logger LOG = LoggerFactory.getLogger(StreamsThreadDelegatingMetricsReporter.class);
```
nit: Generally Kafka code declares logger as `log`, I am not sure about streams code.

```suggestion
    private static final Logger log = LoggerFactory.getLogger(StreamsThreadDelegatingMetricsReporter.class);
```

apoorvmittal10 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -473,6 +474,10 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
         referenceContainer.mainConsumer = mainConsumer;
 
+        final String stateUpdaterId = stateUpdaterEnabled ? threadId.replace("-StreamThread-", "-StateUpdater-") : "NA";
+        final StreamsThreadDelegatingMetricsReporter reporter = new StreamsThreadDelegatingMetricsReporter(mainConsumer, threadId, stateUpdaterId);
```
Do we need `threadId` specific `StreamsThreadDelegatingMetricsReporter` or it can be singleton?

bbejeck replied:
The short answer is that we need a `StreamsThreadDelegatingMetricsReporter` per stream thread. 

For more context, Kafka Streams has an embedded consumer for each `StreamThread` instance. Since we want each embedded consumer to register/deregister metrics only for its "host" StreamThread, we need a reporter instance per thread.

apoorvmittal10 commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,377 @@
+ */
+
+package org.apache.kafka.streams.integration;
+
```
nit: remove additional line break.

apoorvmittal10 reviewed (changes requested):
Thanks for the PR @bbejeck. I have taken a pass and some comments.

bbejeck committed:
Changes per comments, resolving merge conflicts from PR breakdown

bbejeck committed:
Close telemetry reporter on admin client close.

bbejeck commented:
Thanks for the comments @apoorvmittal10 - I've addressed your comments.

bbejeck committed:
Cleanup, revert closing telemetry reporter pending investigation, remove setting consumer to null in reporter close

bbejeck committed:
Checkstyle errors

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1823,6 +1824,26 @@ default ListShareGroupsResult listShareGroups() {
+     * that are available for subscription and sent as
+     * telemetry data to the broker.
+     *
+     * @param metric, the application metric to register
```
```suggestion
     * @param metric The application metric to register.
```

mjsax replied:
Should we talk about error cases? Metric type not supported? What happens if a metric name does already exist (I would assume we update it)?

Same for "remove" below.

Also applies to consumer and producer.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1823,6 +1824,26 @@ default ListShareGroupsResult listShareGroups() {
+     * This metric is removed from this client's metrics
+     * and will not be available for subscription.
+     *
+     * @param metric, the application metric to remove
```
```suggestion
     * @param metric The application metric to remove.
```

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1823,6 +1824,26 @@ default ListShareGroupsResult listShareGroups() {
+    /**
+     * An application to be removed from subscription.
+     * This metric is removed from this client's metrics
+     * and will not be available for subscription.
```
```suggestion
     * and will not be available for subscription any longer.
```

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -556,10 +562,13 @@ static KafkaAdminClient createInternal(
-                (hostResolver == null) ? new DefaultHostResolver() : hostResolver);
+                (hostResolver == null) ? new DefaultHostResolver() : hostResolver,
+                    null,
+                    clientTelemetryReporter.map(ClientTelemetryReporter::telemetrySender).orElse(null));
```
nit: is the indention of the two lines above correct?

mjsax commented on clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java:
```diff
@@ -77,6 +78,8 @@ public class MockConsumer<K, V> implements Consumer<K, V> {
     private Uuid clientInstanceId;
     private int injectTimeoutExceptionCounter;
 
+    final List<KafkaMetric> addedMetrics = new ArrayList<>();
```
Should is be `private` (given that we add `addedMetrics()` method) ?

bbejeck replied:
missed this, updated

mjsax commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadDelegatingMetricsReporter.java:
```diff
@@ -0,0 +1,86 @@
+    
+    private static final Logger log = LoggerFactory.getLogger(StreamsThreadDelegatingMetricsReporter.class);
+    private static final String THREAD_ID_TAG = "thread-id";
+    private Consumer<?, ?> consumer;
```
```suggestion
    private Consumer<byte[], byte[]> consumer;
```

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,376 @@
+    @ParameterizedTest
+    @MethodSource("singleAndMultiTaskParameters")
+    @DisplayName("Streams metrics should get passed to Consumer")
+    void shouldPassMetrics(final String topologyType, final boolean stateUpdaterEnabled) throws InterruptedException {
```
Why is this not `public` -- I thought that a requirement? (also below)

mjsax replied:
Is it worth to make the `stateUpdaterEnabled` distinction? (Not 100% sure what difference it would make and why we need to test it expliclity?)

bbejeck replied:
> Why is this not public -- I thought that a requirement? 

I seem to recall IntelliJ flagging them for being `public`, but I could be mistaken - I've updated them to be so

bbejeck replied:
> Is it worth to make the stateUpdaterEnabled distinction? (Not 100% sure what difference it would make and why we need to test it expliclity?)

In a word, yes.  Having the `stateupdater` enabled creates differently tagged metrics we need to account for, using the test with these parameters I caught a couple of bugs in registering metrics, so I'll keep this as is.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,376 @@
+
+        @Override
+        public Admin getAdmin(final Map<String, Object> config) {
+            return AdminClient.create(config);
```
Do we also need to intercept the admin client to verify the instance level metric we sent using admin?

mjsax reviewed (commented):
Not sure if I understand the end-to-end wiring already, so need to make another pass.

Btw: should we also verify state store metrics (or are they automatically verified as part of task metrics)?

bbejeck committed:
Address review comments: More descriptive javadoc

bbejeck committed:
Address review comments: indentation, consumer generic types

bbejeck committed:
Address review comments: update test to account for correct state metrics, change visibility on test methods to public

bbejeck commented:
> should we also verify state store metrics (or are they automatically verified as part of task metrics)?

They are covered as part of task metrics, but I expanded the test to explicitly verify state store metrics are correct as well

bbejeck committed:
Merge  Admin delegate for testing sending metrics via admin client

bbejeck committed:
Rebasing work

Added a new class `StreamsThreadMetricsDelegatingReporter` to handle metrics reporting for Kafka Streams threads. Updated method names across multiple files to improve clarity and consistency by changing `registerMetric` to `registerMetricForSubscription` and `unregisterMetric` to `unregisterMetricFromSubscription`. Additionally, included new tests to verify the changes.

bbejeck committed:
Refactor and add telemetry support in integration test

Refactored `KafkaStreamsTelemetryIntegrationTest` to use an added a custom telemetry reporter `TestingClientTelemetry` to integrate telemetry with the test environment. Updated checkstyle to permit new telemetry-related imports.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1823,6 +1824,35 @@ default ListShareGroupsResult listShareGroups() {
+     * - `Sum`: Monotonic total count meter (Counter). Suitable for metrics like total number of X, e.g., total bytes sent.
+     * - `Gauge`: Non-monotonic current value meter (UpDownCounter). Suitable for metrics like current value of Y, e.g., current queue count.
+     * Metrics not matching these types are silently ignored.
+     * Executing this method for a previously registered metric is a benign operation and results in updating that metrics entry.
```
> benign

I just learned a new word :)

apoorvmittal10 replied:
Same here.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1823,6 +1824,35 @@ default ListShareGroupsResult listShareGroups() {
+     * The provided metric must map to an OTLP metric data point
+     * type in the OpenTelemetry v1 metrics protobuf message types.
+     * Specifically, the metric should be one of the following:
+     * - `Sum`: Monotonic total count meter (Counter). Suitable for metrics like total number of X, e.g., total bytes sent.
```
We need to us html markup here to get a proper bullet point list -- it won't render as bullet points as-is.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1823,6 +1824,35 @@ default ListShareGroupsResult listShareGroups() {
 
+
+    /**
+     * An application metric provided for subscription.
```
Sound more like a description of the parameter than the method?

Maybe:
```
Add the provided application metric for subscription.
```

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1823,6 +1824,35 @@ default ListShareGroupsResult listShareGroups() {
+    void registerMetricForSubscription(KafkaMetric metric);
+
+    /**
+     * An application to be removed from subscription.
```
Similar to `register` method (also `application [metric]`)

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1823,6 +1824,35 @@ default ListShareGroupsResult listShareGroups() {
+     * This metric is removed from this client's metrics
+     * and will not be available for subscription any longer.
+     * Executing this method with a metric that has not been registered is a
+     * benign operation and does not result in any action taken (no-op)
```
```suggestion
     * benign operation and does not result in any action taken (no-op).
```

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -556,10 +562,13 @@ static KafkaAdminClient createInternal(
                 time,
                 1,
                 (int) TimeUnit.HOURS.toMillis(1),
+        null,
```
nit: indention

bbejeck replied:
renders lined up in Intellij, but I've tried to adjust it.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -556,10 +562,13 @@ static KafkaAdminClient createInternal(
                 metadataManager.updater(),
-                (hostResolver == null) ? new DefaultHostResolver() : hostResolver);
+                (hostResolver == null) ? new DefaultHostResolver() : hostResolver,
+ null,
```
indention

bbejeck replied:
same as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsClientMetricsDelegatingReporter.java:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+
+    boolean isStreamsClientMetric(final KafkaMetric metric) {
```
Should this be `private` ?

bbejeck replied:
Yes - missed this somehow

mjsax commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadMetricsDelegatingReporter.java:
```diff
@@ -0,0 +1,86 @@
+        }
+    }
+
+    boolean tagMatchStreamOrStateUpdaterThreadId(final KafkaMetric metric) {
```
`private` ?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,474 @@
+
+    @Test
+    @DisplayName("Calling unregisterMetric on metrics not registered should not cause an error")
+    public void shouldNotThrowExceptionWhenRemovingNonExistingMetrics() throws InterruptedException {
```
Do we need to test this with a heavy-weight integration test? Or could we unit-test this in a more light weight fashion?

bbejeck replied:
I don't think so.  I used this test with a "live" consumer because the call to remove metrics goes into the `ClientTelemetryReporter`, and I want to confirm no exceptions at this level. Mocking would miss this interaction.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,474 @@
+
+    @ParameterizedTest
+    @MethodSource("singleAndMultiTaskParameters")
+    @DisplayName("Streams metrics should get passed to Consumer")
```
Similar to above? Should we add this to `StreamThreadTest` as unit-test?

bbejeck replied:
Possibly, but the `KafkaStreams` constructor creates the admin client and the `Reporter` instance to pass streams client metrics at this point as well.  Given the integration test is needed anyway to verify metrics track with task assignment changes due to a rebalance, it isn't a burden to have this in an integration test.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,474 @@
+    }
+
+    @Test
+    @DisplayName("Streams metrics should not be visible in consumer metrics")
```
Sounds unit-testable via `StreamThreadTest` ?

bbejeck replied:
I'll leave it as is for the same reason as above.  This test asserts that passing stream metrics to the consumer, which in turn adds them the `ClientTelemetryReporter` does not bleed over to the JMX reporter.

mjsax reviewed (commented):
Thanks for the PR update. Made another pass.

bbejeck committed:
Updates per review comments

bbejeck commented:
@mjsax thanks for the second review, I've addressed your comments

apoorvmittal10 commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsClientMetricsDelegatingReporter.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    private boolean isStreamsClientMetric(final KafkaMetric metric) {
+        final boolean shouldInclude = metric.metricName().group().equals("stream-metrics");
```
What about `stream-thread-metrics`, `stream-task-metrics`, `stream-processor-node-metrics`, `stream-state-metrics` and `stream-record-cache-metrics`? Are they not needed or they fall in same `stream-metrics` group?

bbejeck replied:
@apoorvmittal10, those groups are not part of client-level grouping and get reported separately

apoorvmittal10 commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsClientMetricsDelegatingReporter.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.Map;
+import java.util.Objects;
+
+public class StreamsClientMetricsDelegatingReporter implements MetricsReporter {
```
And are we going to expect this reporter being registered additionally in `metrics.reporter` server config to send Streams metrics?

apoorvmittal10 replied:
Ahhh, I see it's added as default metrics reporter on metrics instance.

bbejeck committed:
Add the telemetry reporter to the reporters list after creation

bbejeck committed:
Fix merge conflic from Fix broken tests in KafkaAdminClientTest

bbejeck committed:
Minor cleanup

bbejeck committed:
Fix merge conflict from Disable metrics push in AdminClient by default; update tests

bbejeck committed:
Fix merge conflict from fix tests, start of end-to-end metrics test

bbejeck committed:
Call initiate close on ClientTelemetryReporter during admin close

bbejeck commented:
@mjsax @apoorvmittal10 comments addressed

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,497 @@
+                    .map(Map.Entry::getValue)
+                    .findFirst().get();
+            assertNotNull(adminInstanceId);
+            assertNotNull(mainConsumerInstanceId);
```
I have more to do with this test, but having it complete should not block the PR progress.

mjsax replied:
Seems it's not completed yet?

bbejeck replied:
Still in flight - but I don't think this should block the PR

bbejeck replied:
Passable implementation complete

AndrewJSchofield commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -474,6 +475,10 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
         taskManager.setMainConsumer(mainConsumer);
         referenceContainer.mainConsumer = mainConsumer;
 
+        final String stateUpdaterId = stateUpdaterEnabled ? threadId.replace("-StreamThread-", "-StateUpdater-") : "NA";
```
This "NA" seems a little inelegant. Philosophically, is there a special state updater with ID "NA", or is there no state updater? I think the latter, in which case I would prefer not to have a fake ID at all.

bbejeck replied:
Fair enough, reflecting some, I'm unsure what I was getting at.  Updated

AndrewJSchofield replied:
Well, I really meant don't pass in a state updater ID at all if there is no state updater enabled, but it's a very small point.

bbejeck replied:
At one point, the state updater was experimental and controlled by an internal config.  I did some follow-up, and as of the 3.8 release, it's enabled so I think we are good here.

AndrewJSchofield replied:
Thanks for looking into it.

AndrewJSchofield reviewed (changes requested):
Just one remaining comment.

bbejeck committed:
Updates per comments

bbejeck commented:
@AndrewJSchofield, thanks for the review; I've addressed your comment.

AndrewJSchofield reviewed (approved):
lgtm

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -556,10 +562,13 @@ static KafkaAdminClient createInternal(
                 time,
                 1,
                 (int) TimeUnit.HOURS.toMillis(1),
+            null,
```
nit: formatting

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -556,10 +562,13 @@ static KafkaAdminClient createInternal(
                 metadataManager.updater(),
-                (hostResolver == null) ? new DefaultHostResolver() : hostResolver);
+                (hostResolver == null) ? new DefaultHostResolver() : hostResolver,
+     null,
```
nit: formatting

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -663,6 +676,8 @@ public void close(Duration timeout) {
         long newHardShutdownTimeMs = now + waitTimeMs;
         long prev = INVALID_SHUTDOWN_TIME;
+        final long reporterTimeoutMs = waitTimeMs;
+        clientTelemetryReporter.ifPresent(reporter -> reporter.initiateClose(reporterTimeoutMs));
```
If this operation can fail, would we want just let it bubble up from `close()` as we do right now, of would it make sense to catch-log-and-swallow to move forward this `close()`?

bbejeck replied:
The `ClientTelemetryReporter.initiateClose` does the catch-log-and-swallow if there's exception in this method

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -691,6 +706,7 @@ public void close(Duration timeout) {
                 // Wait for the thread to be joined.
                 thread.join(waitTimeMs);
             }
+            metrics.close();
```
Why are we doing this inside this try-catch-block?

bbejeck replied:
Oversight on my part, moving it to above the try-catch

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1823,6 +1824,41 @@ default ListShareGroupsResult listShareGroups() {
+     *
+     * @param metric The application metric to register
+     */
+    void registerMetricForSubscription(KafkaMetric metric);
```
Are existing client metrics isolated from app metrics? Or could it happen that a app metric would overwrite an existing one? I could not find any guard?

Similar for `unregisterMetricFromSubscription` below: could this delete a client metric? Could not find a guard either.

If both are not isolated, we might want to put a guard in place, and let each method throw an exception if an existing client metric would be messed with? (And update the JavaDocs accordingly)

mjsax replied:
I was just re-reading the KIP discussion thread, and there was a comment from @lucasbru to use `Metric` interface instead of `KafkaMetric` class as parameter. The reply was, "yes will update the KIP" but seem this never happened?

Btw: cf my other commend below, about checking metric type...

bbejeck replied:
Yes, but @lucasbru's comment was related to changin `Collection<Kafka Metric>` to `Collection<? extends Metric>`  but since the KIP changed to accept a single metric, it makes sense to use `KafkaMetric`.  That approach is consistent with the [Metrics.registerMetric](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java#L589) and the [MetricsReporter.metricChange](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/metrics/MetricsReporter.java#L45) methods.

mjsax replied:
Seems you missed my first qustions:

> Are existing client metrics isolated from app metrics? Or could it happen that a app metric would overwrite an existing one? I could not find any guard?
>
> Similar for unregisterMetricFromSubscription below: could this delete a client metric? Could not find a guard either.
> 
> If both are not isolated, we might want to put a guard in place, and let each method throw an exception if an existing client metric would be messed with? (And update the JavaDocs accordingly)

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4134,6 +4150,22 @@ private KafkaFutureImpl<List<MemberIdentity>> getMembersFromGroup(String groupId
+    public void registerMetricForSubscription(KafkaMetric metric) {
+        if (clientTelemetryReporter.isPresent()) {
+            ClientTelemetryReporter reporter = clientTelemetryReporter.get();
+            reporter.metricChange(metric);
```
Where do we check the `KafkaMetric` type, ie, OTEL compatibility and log a WARN? Seems we would need to do this here before adding it to the `reporter`?

Even if I am not 100% sure how to implement this right now? Could we use `KafkaMetric#MetricValueProvider` and enforce that only certain sub-classes like `Gauge` are supported?

For this case, using `Metric` interface might not even work, as it does not expose any type information (cf my comment above about `KafkaMetric` vs `Metric` in the public interface)?

If we cannot implement this guard, we should update the KIP? Also, what happens effectively? If we would register a `String` metric, the registration itself might not do any harm? But what happens when the broker starts to subscribe to this metic? Would the subscription be ignored (and we log an error when we try to send the metric to the broker, or in the worst case, even crash the client)?

For either case, we need to add tests...

bbejeck replied:
I don't feel this is an issue, for the following reason.  
The `ClientTelemetryReporter.metricChange` delegates the metric to the `KafkaMetricsCollector`. In the [KafkaMetricsCollector.collectMetric](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/telemetry/internals/KafkaMetricsCollector.java#L202-L220) method, it evaluates each metric to ensure it is the expected type. If it's no
a log.debug is issued, and the metric is dropped.

mjsax commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -470,6 +471,19 @@ public void testExplicitlyEnableJmxReporter() {
     }
 
+    @Test
+    public void testExplicitlyEnableTelemetryReporter() {
```
Should we add a "negative test" to ensure that the telementry-reported is not added by default?

mjsax replied:
Did you miss this one?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -993,12 +995,17 @@ private KafkaStreams(final TopologyMetadata topologyMetadata,
         this.clientSupplier = clientSupplier;
-        adminClient = clientSupplier.getAdmin(applicationConfigs.getAdminConfigs(ClientUtils.adminClientId(clientId)));
+        final Map<String, Object> adminConfigs = applicationConfigs.getAdminConfigs(ClientUtils.adminClientId(clientId));
+        adminConfigs.put(AdminClientConfig.ENABLE_METRICS_PUSH_CONFIG, true);
```
I think we should do this inside `getAdminConfigs(...)` and add `ADMIN_DEFAULT_OVERRIDES` similar to what we have for producer/consumer to `StreamsConfig`.

Plus add new test to `StreamsConfigTest` to cover the different edge cases, as describe on the KIP: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=315494236#KIP1076:MetricsforclientapplicationsKIP714extension-Configuration

bbejeck replied:
Done via #17422

mjsax commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsClientMetricsDelegatingReporter.java:
```diff
@@ -0,0 +1,79 @@
+    private boolean isStreamsClientMetric(final KafkaMetric metric) {
+        final boolean shouldInclude = metric.metricName().group().equals("stream-metrics");
+        if (!shouldInclude) {
+            log.debug("Rejecting thread metric {}", metric.metricName());
```
Is this useful for debug level, or should we even push this down to TRACE?

bbejeck replied:
Resolved via #17377

mjsax commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadMetricsDelegatingReporter.java:
```diff
@@ -0,0 +1,86 @@
+        final Map<String, String> tags = metric.metricName().tags();
+        final boolean shouldInclude = tags.containsKey(THREAD_ID_TAG) && (tags.get(THREAD_ID_TAG).equals(threadId) || tags.get(THREAD_ID_TAG).equals(stateUpdaterThreadId));
+        if (!shouldInclude) {
+            log.warn("Rejecting metric {}", metric.metricName());
```
This should not be WARN (we expect that we reject metrics, but should avoid such warn noise) but DEBUG (or even TRACE? -- cf. comment above)

bbejeck replied:
Fixed via #17376

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -474,6 +475,10 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
         taskManager.setMainConsumer(mainConsumer);
         referenceContainer.mainConsumer = mainConsumer;
 
+        final String stateUpdaterId = threadId.replace("-StreamThread-", "-StateUpdater-");
```
Might be good to extract `private final static String ...` vars for both string? So far, each one only appears ones, but now that each appears twice it might be good to use a constant instead of hardcoded on two different places?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,497 @@
+            final ClientInstanceIds clientInstanceIds = streams.clientInstanceIds(Duration.ofSeconds(60));
+            final Uuid adminInstanceId = clientInstanceIds.adminInstanceId();
+            final Uuid mainConsumerInstanceId = clientInstanceIds.consumerInstanceIds().entrySet().stream()
+                    .filter(entry -> entry.getKey().endsWith("1-consumer"))
```
`1-consumer` ?

bbejeck replied:
This was done to get the main consumer and not the restore consumer - I'll update the filter to be
```
 .filter(entry -> !entry.getKey().contains("restore"))
```

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,497 @@
+
+    @ParameterizedTest
+    @MethodSource("singleAndMultiTaskParameters")
+    @DisplayName("Streams metrics should get passed to Consumer")
```
```suggestion
    @DisplayName("Streams metrics should get passed to Admin and Consumer")
```

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,497 @@
+        final Properties properties = props(stateUpdaterEnabled);
+        final Topology topology = topologyType.equals("simple") ? simpleTopology() : complexTopology();
+        /*
+           This test verifies that all Kafka Streams metrics with a thread-id tag get passed to the consumer
```
Seems comment missed `Admin` -- also: does this comment add value to begin with, or should we just remove it entirely?

bbejeck replied:
ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,497 @@
+                    .filter(metricName -> metricName.tags().containsKey("thread-id")).collect(Collectors.toList());
+
+            final List<MetricName> streamsClientMetrics = streams.metrics().values().stream().map(Metric::metricName)
+                    .filter(metricName -> metricName.group().equals("stream-metrics")).collect(Collectors.toList());
```
Seems we get _all_ client-level metrics here, including stuff which is not OTEL compatible? Sounds like a problem to me?

bbejeck replied:
As stated [above](https://github.com/apache/kafka/pull/17021#discussion_r1804849291) I'm not at all convinced that it presents a problem.  In the [KafkaMetricsCollector.collectMetric](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/telemetry/internals/KafkaMetricsCollector.java#L202-L220) method, it evaluates each metric to ensure it is the expected type.  If it's no 
 a `log.debug` is issued, and the metric is dropped.

If you would prefer, I can perform a similar type check on metrics before passing them through, but since it's handled in the `KafkaMetricsCollector` I don't think it's necessary.

mjsax replied:
I see -- given that it only result in DEBUG level logs should be fine. -- If it would result in WARN level logs it could be confusing to users and adding a filter and only pass in valid metrics would be good. But seems, we don't need to worry about it.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,497 @@
+        properties2.put(StreamsConfig.CLIENT_ID_CONFIG, appId + "-ks2");
+
+        /*
+          This test ensures metrics are registered and removed correctly with Kafka Steams dynamic membership changes
```
What is the value of this comment? Seem `@DisplayName` describe it already?

bbejeck replied:
removed

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -1812,7 +1812,6 @@ public Map<String, Object> getAdminConfigs(final String clientId) {
 
         // add client id with stream client id prefix
         props.put(CommonClientConfigs.CLIENT_ID_CONFIG, clientId);
-
```
?

bbejeck replied:
blank line should be back in now

mjsax replied:
Still see it :)

bbejeck committed:
Fix missed items during rebase

bbejeck committed:
Fix constructor that got mangled during rebase

apoorvmittal10 commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClientConfig.java:
```diff
@@ -191,7 +191,7 @@ public class AdminClientConfig extends AbstractConfig {
                                 .define(ENABLE_METRICS_PUSH_CONFIG,
                                         Type.BOOLEAN,
-                                        true,
+                                        false,
```
👍

apoorvmittal10 reviewed (approved):
Thanks for the PR, LGTM!

bbejeck committed:
Fixes for resolve conflict merge mistakes during rebase

bbejeck committed:
Address review comments

bbejeck committed:
Checkstyle fixes

bbejeck committed:
Address Comments

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1844,7 +1844,7 @@ default ListShareGroupsResult listShareGroups() {
      * Executing this method for a previously registered metric is a benign operation and results in updating that metrics entry.
      *
-     * @param metric The application metric to register
+     * @param metric, the application metric to register
```
Seems like a rebase error? Original code looks right

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -556,11 +561,12 @@ static KafkaAdminClient createInternal(
                 time,
                 1,
-                (int) TimeUnit.HOURS.toMillis(1),
+                (int) TimeUnit.HOURS.toMillis(1), null,
```
nit: `null` should go into it's own line

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -556,11 +561,12 @@ static KafkaAdminClient createInternal(
+                (int) TimeUnit.HOURS.toMillis(1), null,
                 metadataManager.updater(),
-                (hostResolver == null) ? new DefaultHostResolver() : hostResolver);
+                (hostResolver == null) ? new DefaultHostResolver() : hostResolver, null,
```
as above

mjsax commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -664,6 +674,8 @@ public void close(Duration timeout) {
         long newHardShutdownTimeMs = now + waitTimeMs;
         long prev = INVALID_SHUTDOWN_TIME;
+        clientTelemetryReporter.ifPresent(ClientTelemetryReporter::initiateClose);
+        metrics.close();
```
Just double checking: could either `initiateClose` or `metrics.close()` throw? If yes, would it be a problem that we don't execute the code below?

bbejeck replied:
No, neither method throws.

mjsax reviewed (commented):
Did not look at test code yet.

bbejeck committed:
Completed integration test

bbejeck committed:
Get initial end-to-end integration test running

bbejeck committed:
Respond to comments

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,533 @@
+        public void exportMetrics(final AuthorizableRequestContext context, final ClientTelemetryPayload payload) {
+            try {
+                final Uuid clientId = payload.clientInstanceId();
+                //Temporary until a solution is found for getting shaded MetricData visible to gradle
+                final int dataSize = payload.data().array().length;
+                SUBSCRIBED_CLIENT_METRIC_DATA_COUNT.put(clientId, dataSize);
+
+            } catch (final Exception e) {
```
~I needed to do this as trying parse the telemetry data with `MetricsData` object from open telemetry is not possible at the moment because gradle can't find the import `org.apache.kafka.shaded.io.opentelemetry.proto.metrics.v1.MetricsData` so until a solution is found I think this is a reasonable proxy for getting a metrics dump.~

bbejeck replied:
This is resolved now.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,533 @@
+
+            TestUtils.waitForCondition(() -> SUBSCRIBED_CLIENT_METRIC_DATA_COUNT.get(mainConsumerInstanceId) >= 10_000,
+                    30_000 * 10, //Temporary until a workaround for getting past the initial push interval of 5 minutes then set at 30 seconds
+                    "Never received subscribed metrics");
```
~As the comment indicates, this is temporary until the metrics push timing can be worked out.~

This has been fixed now.

bbejeck committed:
More descriptive name

bbejeck committed:
Reslove conflict from this commit Add open telemetry protobuf as test dependency, update import control, fix test

bbejeck commented on build.gradle:
```diff
@@ -2638,6 +2638,7 @@ project(':streams') {
     testImplementation libs.mockitoCore
     testImplementation libs.mockitoJunitJupiter // supports MockitoExtension
     testImplementation libs.junitPlatformSuiteEngine // supports suite test
+    testImplementation libs.opentelemetryProto
```
Needed to add this test dependency explicitly.  The clients module has this dependency shaded, but it's the package is renamed from `io.opentelemetry.proto` to  `org.apache.kafka.shaded.io.opentelemetry.proto`. IntelliJ finds this fine, but my local gradle build can't so I need to add it.

mjsax replied:
@apoorvmittal10 Thoughts? I am not a gradle expert... I am fine with this. Just checking if you have any thoughts?

apoorvmittal10 replied:
@bbejeck This doesn't seems right.

We should use `org.apache.kafka.shaded.io.opentelemetry.proto.metrics.v1.MetricsData` as dependency in `KafkaStreamsTelemetryIntegrationTest`.  And change `streams` project dependency in build.gradle from `api project(':clients')` to `api project(path: ':clients', configuration: 'shadow')`.

I have verified the change by taking your PR and applying the fixes mentioned above. You need to amend import-control to allow dependency.

Also prease verify the contents of the streams jar as well, regarding what all is included. You can do so by issuing `jar -tf kafka-streams.jar  > jar_tf_output` (pre and post). Also check the manifest file.

bbejeck replied:
Thanks, @apoorvmittal10 this definitely seems better I've updated it.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,547 @@
+                    30_000,
+                    "Never received subscribed metrics");
+            final List<String> actualTaskMetrics = SUBSCRIBED_CLIENT_METRICS.get(mainConsumerInstanceId).stream().filter(metricName -> metricName.startsWith("org.apache.kafka.stream.task")).collect(Collectors.toList());
+            assertEquals(EXPECTED_MAIN_CONSUMER_TASK_METRICS, actualTaskMetrics);
```
Picking grabbing task metrics as a good indicator that streams metrics push is working.

mjsax replied:
Given the test name, can't we access the broker side plugin to verify what metrics got pushed? Should we verify instance-level, ie, `admin`, too?

bbejeck replied:
> Given the test name, can't we access the broker side plugin to verify what metrics got pushed?

That's exactly what the test does, the Map containing the metrics name is populated in the plugin but this may have been obscured a bit by naming - I've updated the code to make it more clear what's going on

bbejeck commented:
@mjsax 

> Did not look at test code yet.

I've updated the integration test and I consider it completed at this point

mjsax commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -469,6 +471,30 @@ public void testExplicitlyEnableJmxReporter() {
+    public void testExplicitlyEnableTelemetryReporter() {
+        Properties props = new Properties();
+        props.setProperty(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        props.setProperty(AdminClientConfig.METRIC_REPORTER_CLASSES_CONFIG, "org.apache.kafka.common.metrics.JmxReporter");
```
Why do we need to set this config?

bbejeck replied:
Don't need to, shamefully probably a c/p error updated

mjsax commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -469,6 +471,30 @@ public void testExplicitlyEnableJmxReporter() {
+    public void testExplicitlyTelemetryReporterIsDisabled() {
+        Properties props = new Properties();
+        props.setProperty(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        props.setProperty(AdminClientConfig.METRIC_REPORTER_CLASSES_CONFIG, "org.apache.kafka.common.metrics.JmxReporter");
```
As above. Can't we test if number of reporters is 0 instead of 1

bbejeck replied:
fixed

mjsax commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -469,6 +471,30 @@ public void testExplicitlyEnableJmxReporter() {
+        assertEquals(2, admin.metrics.reporters().size());
+        //ClientTelemetryReporter always added after metrics reporters created with JmxReporter
+        assertInstanceOf(ClientTelemetryReporter.class, admin.metrics.reporters().get(1));
+        admin.close();
```
Should we use try-with-resource when creating admin instead of explicit close?

mjsax commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -469,6 +471,30 @@ public void testExplicitlyEnableJmxReporter() {
+        KafkaAdminClient admin = (KafkaAdminClient) AdminClient.create(props);
+        assertEquals(1, admin.metrics.reporters().size());
+        assertInstanceOf(JmxReporter.class, admin.metrics.reporters().get(0));
+        admin.close();
```
as above

mjsax commented on core/src/test/java/kafka/admin/ClientTelemetryTest.java:
```diff
@@ -80,7 +80,7 @@ public class ClientTelemetryTest {
         Map<String, Object> configs = new HashMap<>();
         configs.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, clusterInstance.bootstrapServers());
-        configs.put(AdminClientConfig.ENABLE_METRICS_PUSH_CONFIG, true);
+        configs.put(AdminClientConfig.ENABLE_METRICS_PUSH_CONFIG, "true");
```
Why this change?

bbejeck replied:
unintentional  - will revert

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,547 @@
+
+    @Test
+    @DisplayName("Calling unregisterMetric on metrics not registered should not cause an error")
+    public void shouldNotThrowExceptionWhenRemovingNonExistingMetrics() throws InterruptedException {
```
Should we have a test (might be a `clients` test though, not a `streams` test) in which we try to remove some **_actual_** client metric (best one test for each client, ie, consumer, producer, and admin), and verify that the underlying client metric was **_not_** removed (ie, test proper isolation)?

This actually also makes we wonder about this test: should it be a `clients` test? In the end, we get `embeddedConsumer` to call `unregisterMetricFromSubscription` -- Why would we need to do this test in the context of KS?

bbejeck replied:
Originally I believed this belongs in the context of KS because KS will be actively removing and adding metrics, but that is handled in the test with dynamic membership changes.  You make a valid point here, but I would argue that could be separate from this PR and done in a separate PR concurrently.  At a minimum, I'll remove the test from this PR.

bbejeck replied:
Created a ticket for this https://issues.apache.org/jira/browse/KAFKA-17869

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,547 @@
+                LOG.info("Found metrics {} for clientId={}", metricNames, clientId);
+                SUBSCRIBED_CLIENT_METRICS.put(clientId, metricNames);
+            } catch (final Exception e) {
+                e.printStackTrace(System.out);
```
Should this to go `System.err` ?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,547 @@
+    public void tearDown() throws Exception {
+        INTERCEPTING_CONSUMERS.clear();
+        INTERCEPTING_ADMIN_CLIENTS.clear();
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfigurations);
```
Given that we execute this with `@AfterEach`, why do we need to use `streamsConfiguration[s]` -- seems unnecessary, and we can just use a single class level `Properties streamsConfig` ?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,547 @@
+
+            final Map<MetricName, ? extends Metric> embeddedConsumerMetrics = INTERCEPTING_CONSUMERS.get(FIRST_INSTANCE_CONSUMER).metrics();
+
+            streamsThreadMetrics.forEach(metricName -> assertFalse(embeddedConsumerMetrics.containsKey(metricName), "Stream thread metric found in client metrics" + metricName));
```
Should we verify this for `admin`, too?

mjsax reviewed (commented):
Should we extend the test, to not only verify task metrics, but also thread metrics? What about DEBUG and TRACE level?

bbejeck committed:
Address comments, fix a potential bug and side cleanup

bbejeck committed:
Store clientInstanceId before returning

bbejeck commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4915,47 +4934,16 @@ public Uuid clientInstanceId(Duration timeout) {
-            throw new KafkaException("Error occurred while fetching client instance id", e);
-        }
-
+        clientInstanceId = ClientTelemetryUtils.fetchClientInstanceId(clientTelemetryReporter.get(), timeout);
```
@apoorvmittal10 The `KafkaStreamsTelemetryIntegrationTest.shouldPushMetricsToBroker` exposed an issue where the admin client instance id returned to Kafka Streams didn't match the one included in the pushed metrics.  To fix this, I updated the `clientInstanceId` method to use the same approach in the `ClassicKafkaConsumer` and the integration test is working as expected now.

apoorvmittal10 replied:
Hmmm, so the previous implementation in admin client sent a new getTelemetryRequest which will anyways get the new client instance as it's a new API call. I think the changes you have done are correct.

apoorvmittal10 replied:
Just note the call with timeout might block the thread for that time so please be cautious while invoking `clientIntsanceId` method. There is timed_waiting signal in client telemetry reporter for fetching client instance id.

bbejeck commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -8073,18 +8103,21 @@ public void testFenceProducers() throws Exception {
+            clientTelemetryReporter.configure(any());
+            mockedCommonClientConfigs.when(() -> CommonClientConfigs.telemetryReporter(anyString(), any())).thenReturn(Optional.of(clientTelemetryReporter));
+            
+            try (AdminClientUnitTestEnv env = mockClientEnv(AdminClientConfig.ENABLE_METRICS_PUSH_CONFIG, "true")) {
```
I used nested try-with-resources to ensure all mocking gets established first before the creation of the admin client, which executes `CommonClientConfigs.telemetryReporter`

bbejeck committed:
Address review comments

bbejeck commented:
> should we extend the test, to not only verify task metrics, but also thread metrics? What about DEBUG and TRACE level?

Sure, but I don't think that should block the PR. Given that TRACE will include all metrics I'd say the end-to-end test should simply extend to set that level to ensure all expected metrics make it through.

bbejeck committed:
Update end-to-end test

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,502 @@
+                    30_000,
+                    "Never received subscribed metrics");
+            final List<String> expectedMetrics = streams.metrics().values().stream().map(Metric::metricName)
+                    .filter(metricName -> metricName.tags().containsKey("thread-id")).map(mn -> {
+                        final String name = mn.name().replace('-', '.');
+                        final String group = mn.group().replace("-metrics", "").replace('-', '.');
+                        return "org.apache.kafka." + group + "." + name;
```
Need to format the Kafka Streams metrics names into the KIP-714 naming format convention.

bbejeck committed:
Update end-to-end test to validate metrics for INFO, DEBUG and TRACE

bbejeck committed:
commit f30f2d03e9064cb9a167fe35b0c5d63470405ce6 puts
all clients in the ClientsInstanceIds in consumerInstanceIds map. This PR makes sure only the main consumer client instance id is selected.

bbejeck committed:
Standardize on lower-case if names ever change

bbejeck committed:
Standardize on lower-case if names ever change

mjsax commented:
@bbejeck Can you rebase this PR to resolve merge conflicts?

bbejeck committed:
Rebased trunk and updated test

bbejeck commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,506 @@
+            TestUtils.waitForCondition(() -> !TelemetryPlugin.SUBSCRIBED_METRICS.get(mainConsumerInstanceId).isEmpty(),
+                    30_000,
+                    "Never received subscribed metrics");
+            final List<String> expectedMetrics = streams.metrics().values().stream().map(Metric::metricName)
+                    .filter(metricName -> metricName.tags().containsKey("thread-id")).map(mn -> {
+                        final String name = mn.name().replace('-', '.');
+                        final String group = mn.group().replace("-metrics", "").replace('-', '.');
+                        return "org.apache.kafka." + group + "." + name;
+                    }).sorted().collect(Collectors.toList());
```
Needed to convert the metric names into the expected KIP-714 naming format

bbejeck commented:
@mjsax rebased on trunk - ready for final review

mjsax commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,506 @@
+            
+            final Uuid mainConsumerInstanceId = clientInstanceIds.consumerInstanceIds().entrySet().stream()
+                    .filter(entry -> !entry.getKey().endsWith("-restore-consumer")
+                            && !entry.getKey().endsWith("GlobalStreamThread"))
```
We don't use `-global-consumer` in the key for the global consumer? 

This sounds like a bug to me -- checking the code, it seems we use `getName() + "-consumer"` inside `StreamThread` while we use `getName()` only for the global case inside `KafkaStreams`.

Out of scope for this PR, but I think we should fix it in a follow up PR (given that it is technically a breaking change, we should think about it twice though... But as we head into 4.0, it might be ok -- however, we should call it out in the upgrade docs and we need a Jira for it)

mjsax commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,506 @@
+                    30_000,
+                    "Never received subscribed metrics");
+            final List<String> actualInstanceMetrics = TelemetryPlugin.SUBSCRIBED_METRICS.get(adminInstanceId);
+            final List<String> expectedInstanceMetrics = Arrays.asList("org.apache.kafka.stream.alive.stream.threads", "org.apache.kafka.stream.failed.stream.threads");
```
Unrelated, but "alive stream thread" seems to be missing in the docs: https://kafka.apache.org/documentation/#kafka_streams_client_monitoring

Would you be interested to do a follow up PR about it?

mjsax commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,506 @@
+                        final String group = mn.group().replace("-metrics", "").replace('-', '.');
+                        return "org.apache.kafka." + group + "." + name;
+                    }).sorted().collect(Collectors.toList());
+            final List<String> actualTaskMetrics = new ArrayList<>(TelemetryPlugin.SUBSCRIBED_METRICS.get(mainConsumerInstanceId));
```
Why "task" metrics?

bbejeck replied:
I never updated the name - it's an oversight

mjsax commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,506 @@
+        }
+    }
+
+    public static class MetricsInterceptingConsumer<K, V> extends KafkaConsumer<K, V> {
```
We use `TestingMetricsInterceptingAdminClient` -- should we name this class `TestingMetricsInterceptingConsumer` to align the names?

mjsax commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,506 @@
+        }
+
+        @Override
+        public void exportMetrics(final AuthorizableRequestContext context, final ClientTelemetryPayload payload) {
```
This method is called when metrics are receive broker side, right? -- I am not sure where we select the metrics we want to collect? `init(...)` is empty, so it seems we subscribe to _all_ metrics? Won't this imply we also collect plain client metrics, not just KS metrics (for this case, why does this test pass, as we should collect not only KS metrics but we test for "equals")?

mjsax commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,506 @@
+
+    @Test
+    @DisplayName("Streams metrics should not be visible in consumer metrics")
+    public void passedMetricsShouldNotLeakIntoConsumerMetrics() throws Exception {
```
```suggestion
    public void passedMetricsShouldNotLeakIntoClientMetrics() throws Exception {
```

mjsax commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,506 @@
+    }
+
+    @Test
+    @DisplayName("Streams metrics should not be visible in consumer metrics")
```
```suggestion
    @DisplayName("Streams metrics should not be visible in client metrics")
```

mjsax commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -0,0 +1,506 @@
+
+
+            final List<MetricName> consumerPassedStreamThreadMetricNames = INTERCEPTING_CONSUMERS.get(FIRST_INSTANCE_CONSUMER).passedMetrics.stream().map(KafkaMetric::metricName).collect(Collectors.toList());
+            final List<MetricName> adminPassedStreamClientMetricNames = INTERCEPTING_ADMIN_CLIENTS.get(FIRST_INSTANCE_CONSUMER).passedMetrics.stream().map(KafkaMetric::metricName).collect(Collectors.toList());
```
`FIRST_INSTANCE_CONSUMER` -> `FIRST_INSTANCE_CLIENT` ?

We use it for consumer and admin

mjsax reviewed (approved):
Only some nits. Overall LGTM. Feel free to merge.

bbejeck committed:
switch back to combined mode - getting not enough resources error

bbejeck commented:
Merged #17021 into trunk
